src/orchestrator-kubernetes/src/lib.rs

```diff
@@ -859,12 +963,38 @@ impl NamespacedOrchestrator for NamespacedKubernetesOrchestrator {
             .clone()
             .into_iter()
             .collect();
-        if let Some(availability_zone) = availability_zone {
-            node_selector.insert(
-                "materialize.cloud/availability-zone".to_string(),
-                availability_zone,
-            );
-        }
+
+        let node_affinity = if let Some(availability_zones) = availability_zones {
+            let selector = NodeSelectorTerm {
+                match_expressions: Some(vec![NodeSelectorRequirement {
+                    key: "materialize.cloud/availability-zone".to_string(),
+                    operator: "In".to_string(),
+                    values: Some(availability_zones),
+                }]),
+                match_fields: None,
+            };
+
+            if scheduling_config.soften_az_affinity {
+                Some(NodeAffinity {
+                    preferred_during_scheduling_ignored_during_execution: Some(vec![
+                        PreferredSchedulingTerm {
+                            preference: selector,
+                            weight: scheduling_config.soften_az_affinity_weight,
+                        },
+                    ]),
+                    required_during_scheduling_ignored_during_execution: None,
+                })
+            } else {
+                Some(NodeAffinity {
+                    preferred_during_scheduling_ignored_during_execution: None,
+                    required_during_scheduling_ignored_during_execution: Some(NodeSelector {
+                        node_selector_terms: vec![selector],
+                    }),
+                })
+            }
+        } else {
+            None
+        };
```

Comment:
Flagging for posterity that I did not review this file because @doy-materialize did.